### ISSUE TITLE
Using JSON for header values

### ISSUE THREAD

**Issue by annevk** [w3c/reporting#127]
I don't think we should do this until there's agreement from the HTTP WG that this is in fact the way to go.

The current approach the HTTP WG seems to be leaning towards is https://tools.ietf.org/html/draft-ietf-httpbis-header-structure.

cc @mnot

**Comment by dcreager** [comment]
We had a discussion about Structured Headers vs JSON during the TAG review in #53.  Consensus was to move forward with JSON.  [This comment](/w3c/reporting/issues/53#issuecomment-403218275) contains the summary — short version is that the `Report-To` header needs a more nested structure than SH currently supports.

**Comment by annevk** [comment]
I'm rather surprised @mnot and @reschke signed off on it, but so be it I guess. I was really hoping we could move to a world where all new headers would use the structured header syntax, but I guess it's custom grammars forever.

**Comment by dcreager** [comment]
I hope our use of JSON isn't seen as an indictment of Structured Headers, because it's definitely not intended that way.  It's just a question of timing.  Reporting first decided to use JSON headers [~3 years ago](/w3c/reporting/commit/64c724c426c7d464c93e6a31d83d0d0debec6080), well before the Structured Headers spec was first created.  I completely agree with you about not wanting to encourage custom grammars.  That's why we went with JSON!  At the time it was the only way to avoid one.

**Comment by mnot** [comment]
@annevk "signed off" is a bit strong, but our intent isn't to shove this down people's throats; if they have good reasons for another approach, by all means use it.

**Comment by annevk** [comment]
@dcreager I guess that's fair, I didn't quite realize how old this is; though it's still only Chrome that's shipping?

This does end up requiring a JSON parser in the networking code, but I guess other things would make one end up there too (Origin Policy comes to mind).

**Comment by dcreager** [comment]
> though it's still only Chrome that's shipping?

Yes, that's right, only Chrome is shipping so far.  Reporting+NEL went live in M69.

> This does end up requiring a JSON parser in the networking code, but I guess other things would make one end up there too (Origin Policy comes to mind).

Yep, that's true.  For Chrome, we already had one available that was easy to link in, though we did perform an extra security audit of the code before launching.  We had considered doing the JSON parsing in a sidecar process for security reasons, but the network code itself is in the process of being moved into a separate service process, and so it looks like that's not going to be necessary.
